### ISSUE TITLE
webpack 5 release plan

### ISSUE THREAD

**Issue by sokra** [webpack/webpack#11406]
TL;DR: release planned for 2020-10-10

After nearly 1 year of beta testing and about 2 years of development we feel that it's time to release webpack 5 soon. As promised we do announce the release at least one month earlier to allow the ecosystem to catch up. To support that we ~will release~ have released a Release Candidate (RC) ~soon-ish in about 1-2 weeks~. Once the RC has been released we try to avoid breaking changes (expect for experimental features, or when absolutely necessary).

We will use the time until the release to:

* [x] Fix bugs
  * [x]  delete x.y.z #10831
* [x] Update the changelog to give a good picture of the changes
* [x] Update the migration guide
* [ ] Create a separate migration guide for plugin developers
* [x] Update the documentation so that at least a majority of webpack 5 features is documented.
* [x] Update important ecosystem components to be webpack 5 compatible
  * [x] webpack-cli
  * [x] webpack-dev-server/middleware
  * [x] mini-css-extract-plugin
* [ ] Write in-deep guides to the new features
* [x] Improve deprecation messages and add better warnings for breaking changes

This is works out great we release webpack 5 at 2020-10-10 because it's a cool date.

Note that while we consider webpack 5 as pretty stable, there are probably still many bugs and problems. Some are expected breaking changes, some are bugs that we missed. If you have problems upgrading make sure to read the migration guide and changelog where we places some hints how to fix these problems and/or why we did some of the breaking changes. If you still run into problems, it might be a bug or something is missing in the migration guide/changelog, please report an issue.

Note that some of the experimental features are still incomplete. They are excluded from SemVer for webpack and we continue to develop them...

Note that minor bugs won't prevent us from releasing. The release is more about finishing breaking changes then about a super stable release. We will release minor/patch version very often to address bugs and add more feature (as we have done with webpack 4).

---

<details><summary><del>We are not yet ready for a release candidate.</del></summary>

The following breaking changes need to be implemented before the RC can be released:

* [x] resolving: `resolve.fallback` and imports field need to be added
  * We consider this a very small breaking change as `imports` field is rarely used and only incorrectly imports field would break
* [x] stats: The stats format and defaults need to be improved for readability (Status: breaking changes done)
  * While mainly affecting the text format, the JSON format might be affected as well.
  * [x] assets grouping
  * [x] modules grouping
  * [x] entrypoints verbosity
  * [x] summary
  * [ ] size color scale
  * [ ] avoid MiB in favor of KiB
  * [ ] 1000s separator in size numbers
  * [ ] hide [emitted]/[built] when all assets/modules shown have it
  * [ ] highlight filename, lowlight path
  * [x] sort assets by size
  * [ ] hint about error details
* [x] assets: The `asset` experiment will be enabled by default.
  * This enables `new URL` parsing in modules and this can break code.
* [x] WebWorker: Add `worker` experiment to handle `new Worker(new URL("./worker", import.meta.url))`, etc.
  * This is not a breaking change, but we might figure out the some internal breaking changes are needed to implement that. That's why it need to be pushed for the RC.
* [x] node.js compat: The `mjs` experiment will be enabled by default. This affects `.mjs` files and `type: "module"` in package.json.
  * Affect files are handled more strict and certain constructs are forbidden. Resolving is also affected (extension is required).
* [x] module concatenation: refactor to make it a first class citizen
* [x] changelog: The changelog is outdated and need to be updated for recent changed (Status: beta.21 -> beta.30 missing)
* [x] improved target option
  * handle in defaulting
  * [x] `externalsPresets`
  * [x] `loader.target`
  * [x] `output.ecmaVersion` -> `output.environment.xxx`
  * [x] `output.environment.global`
* [x] enhanced-resolve, loader-runner and webpack-sources from beta -> stable
* [x] `output.publicPath: "auto"` as default?
* Known bugs:
  * [ ] Module Concatenation: `delete a.b.c` is broken (Status: Parsing need to be changed a bit)
  * [x] Module Federation: Circular remotes are broken (Stats: requires a breaking change on the API)
* (list might be incomplete...)

</details>

---

Known Problems blocking release:

* [x] `delete x.y.z` doesn't work with `optimization.concatenateModules: true` yet.
* [x] `mini-css-extract-plugin` is not fully compatible and there a few problems.
* [x] `html-webpack-plugin` doesn't understand the new default automatic publicPath yet. Use `output.publicPath: ""` instead.
* [x] `target` doesn't support individual browser versions yet. Use the general targets for now: `target: ["web", "es2020"]`
* [x] The stable `webpack-cli` shows too verbose output for schema validation problems.
* [x] `new URL` with string not starting with `./` or `../` works incorrectly
* [x] The stats grouping algorithm need still some fine-tuning

---

Please don't report any outstanding issues in this thread. Please create a new issue (Reference this issue)!

This thread is reserved for feedback how happy you are that webpack 5 is released, to post positive experience with webpack 5 if you already tried the beta and to post any announcements that ecosystem loaders/plugins/tools are webpack 5 compatible.

I'll update the issue with more information when available.

**Comment by ycjcl868** [comment]
awesoming ~

**Comment by artem-malko** [comment]
As for me, webpack 5 is super-fast against its 4 version. Built-in cache works perfectly) And it wasn't to hard to switch from webpack 4 to webpack 5! 

There are some issues, how to migrate my own loaders and plugins, but there are just some warnings, and it is nice, cause you've made so many cool features, and there is no serious problems with version switching.

**Comment by tenjojeremy** [comment]
Amazing work webpack team 👍

**Comment by syuilo** [comment]
Thank you for your hardwork🥰

**Comment by davidcalhoun** [comment]
Thank you!  Looking forward to this, especially [the ability to output modules](https://github.com/webpack/webpack/issues/2933), which is why I'm using Rollup at the moment.

BTW: 2020-10-10 = 👓-🙌- 🙌

**Comment by Linkontoask** [comment]
Look forward to!

**Comment by k644606347** [comment]
Thank you for your hardwork:smiling\_face\_with\_three\_hearts:

**Comment by evilebottnawi** [comment]
@isrmicha What is a problem?

**Comment by csvan** [comment]
Module Federation is going to be a gamechanger for us. Very much looking forward to it!

**Comment by AlonMiz** [comment]
GREAT NEW
issue:
from 5.0.0-beta.27 to 5.0.0-beta.28 i get a peculiar source map error
```
[compile:ci] (node:78) [DEP_WEBPACK_MAIN_TEMPLATE_GET_ASSET_PATH] DeprecationWarning: MainTemplate.getAssetPath is deprecated (use Compilation.getAssetPath instead)
[compile:ci] (node:78) UnhandledPromiseRejectionWarning: Error: "." is not in the SourceMap.
[compile:ci]     at BasicSourceMapConsumer.SourceMapConsumer_sourceContentFor [as sourceContentFor] (/src/node_modules/source-map/lib/source-map-consumer.js:753:13)
[compile:ci]     at /src/app-frontend/node_modules/webpack-sources/lib/applySourceMap.js:144:46
[compile:ci]     at SourceNode_walk [as walk] (/src/node_modules/source-map/lib/source-node.js:230:9)
[compile:ci]     at SourceNode_walk [as walk] (/src/node_modules/source-map/lib/source-node.js:226:13)
[compile:ci]     at applySourceMap (/src/app-frontend/node_modules/webpack-sources/lib/applySourceMap.js:58:13)
[compile:ci]     at SourceMapSource.node (/src/app-frontend/node_modules/webpack-sources/lib/SourceMapSource.js:198:11)
[compile:ci]     at exports.getSourceAndMap (/src/app-frontend/node_modules/webpack-sources/lib/helpers.js:17:21)
[compile:ci]     at SourceMapSource.sourceAndMap (/src/app-frontend/node_modules/webpack-sources/lib/SourceMapSource.js:184:10)
[compile:ci]     at getTaskForFile (/src/app-frontend/node_modules/webpack/lib/SourceMapDevToolPlugin.js:77:30)
[compile:ci]     at /src/app-frontend/node_modules/webpack/lib/SourceMapDevToolPlugin.js:264:22
[compile:ci]     at /src/app-frontend/node_modules/webpack/lib/Cache.js:93:5
[compile:ci]     at Hook.eval [as callAsync] (eval at create (/src/app-frontend/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
[compile:ci]     at Cache.get (/src/app-frontend/node_modules/webpack/lib/Cache.js:75:18)
[compile:ci]     at ItemCacheFacade.get (/src/app-frontend/node_modules/webpack/lib/CacheFacade.js:116:15)
[compile:ci]     at /src/app-frontend/node_modules/webpack/lib/SourceMapDevToolPlugin.js:215:18
[compile:ci]     at arrayEach (/src/node_modules/neo-async/async.js:2405:9)
[compile:ci]     at Object.each (/src/node_modules/neo-async/async.js:2846:9)
[compile:ci]     at /src/app-frontend/node_modules/webpack/lib/SourceMapDevToolPlugin.js:202:15
[compile:ci]     at _next1 (eval at create (/src/app-frontend/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:22:1)
[compile:ci]     at eval (eval at create (/src/app-frontend/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:38:1)
[compile:ci]     at processTicksAndRejections (internal/process/task_queues.js:93:5)
[compile:ci] (node:78) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
[compile:ci] (node:78) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by sokra** [comment]
@AlonMiz please create an issue

**Comment by kunfan96** [comment]
666

**Comment by TheCodingGorilla** [comment]
Really looking forward to Module Federation. Thank you for all yours and the teams hard work. 🙇

**Comment by smoliakov** [comment]
I am sure Module Federation will change microfrontends, looking forward to using it! 
It`s a geat work!

**Comment by sokra** [comment]
https://github.com/webpack/webpack/releases/tag/v5.0.0-rc.0

**Comment by jantimon** [comment]
> `html-webpack-plugin` doesn't understand the new default automatic publicPath yet. Use output.publicPath: "" instead.

Thanks for the heads up ( https://github.com/jantimon/html-webpack-plugin/issues/1514 ) 👍

Is there any api which returns the calculated publicPath?  
Right now we are using https://github.com/jantimon/html-webpack-plugin/blob/f3ccdd5125566ba8d2965cfb4f59b058854fa3c9/index.js#L533-L540

**Comment by fwh1990** [comment]
Who will be faster and more modern between webpack5 and vite?

**Comment by artem-malko** [comment]
> Who will be faster and more modern between webpack5 and vite?

When Vite will work with any code, not only with Vue3)

**Comment by fwh1990** [comment]
> > Who will be faster and more modern between webpack5 and vite?
> 
> Webpack works with any code, not only with Vue3 =)

Vite also works with react and other frameworks.

> Although Vite is primarily designed to work with Vue 3, it can support other frameworks as well. For example, try `npm init vite-app --template react` or `--template preact`.

BTW, webpack is the best tool now.

**Comment by fwh1990** [comment]
> > Who will be faster and more modern between webpack5 and vite?
> 
> Webpack works with any code, not only with Vue3 =)

Vite also works with react and other frameworks.

> Although Vite is primarily designed to work with Vue 3, it can support other frameworks as well. For example, try `npm init vite-app --template react` or `--template preact`.

BTW, webpack is the best tool now.

**Comment by udayrajMT** [comment]
Thank you for making this game changer tool. Hats off for your hard work! 👏👏

**Comment by csvan** [comment]
Are we still on? :)

**Comment by sokra** [comment]
yep. in about 9 hours...

**Comment by csvan** [comment]
Woop!

https://webpack.js.org/blog/2020-10-10-webpack-5-release/
